[Mappings editor] Stack revealed controls vertically

cjcenizal opened the pull request:
Based on discussion with @mdefazio, this changes the layout of revealed controls to stack vertically in order to provide a predictable pattern to the form content and to prevent fields from becoming too cramped horizontally. This PR also adds labels to all fields, even if there is only a single field within a form section, and also refines the copy in the range labels for "Fielddata" and "Index prefixes" and the "Index and search analyzer" label.

# Text

![image](https://user-images.githubusercontent.com/1238659/70097641-b01f8480-15de-11ea-83b0-235f5bfa48d8.png)

![image](https://user-images.githubusercontent.com/1238659/70098035-b2cea980-15df-11ea-8b14-09dd79606485.png)

![image](https://user-images.githubusercontent.com/1238659/70098039-b5310380-15df-11ea-980f-bebb7103f290.png)

![image](https://user-images.githubusercontent.com/1238659/70097672-bc0b4680-15de-11ea-8c51-326c4afa34a2.png)

# Keyword

![image](https://user-images.githubusercontent.com/1238659/70164766-992c7100-1676-11ea-9e13-23e13bddfdca.png)

# Numeric

![image](https://user-images.githubusercontent.com/1238659/70164757-9598ea00-1676-11ea-8f9c-cf0a59b7c3c9.png)

# Boolean

![image](https://user-images.githubusercontent.com/1238659/70165265-7189d880-1677-11ea-993e-a2bab81912de.png)

# Date

![image](https://user-images.githubusercontent.com/1238659/70170602-d8ac8a80-1681-11ea-922d-40fb03804352.png)

# Geo-point

![image](https://user-images.githubusercontent.com/1238659/70180278-888af380-1694-11ea-8b9c-53e1db394565.png)

# Geo-shape

![image](https://user-images.githubusercontent.com/1238659/70180316-9e001d80-1694-11ea-95c0-3a0b8f7e76cc.png)

# Alias

![image](https://user-images.githubusercontent.com/1238659/70181473-ca1c9e00-1696-11ea-86ca-dd9fa857c539.png)
 
# Completion suggester

![image](https://user-images.githubusercontent.com/1238659/70184700-369a9b80-169d-11ea-955c-b87b13e19277.png)

# Flattened

![image](https://user-images.githubusercontent.com/1238659/70184845-8d07da00-169d-11ea-89f8-0c7790010796.png)

# IP

![image](https://user-images.githubusercontent.com/1238659/70188135-c7c14080-16a4-11ea-997b-197d4e60bea4.png)

# Search-as-you-type

![image](https://user-images.githubusercontent.com/1238659/70186175-29cb7700-16a0-11ea-92dc-e2cd69db2b80.png)

![image](https://user-images.githubusercontent.com/1238659/70187884-2afea300-16a4-11ea-8161-e00eeb68f0f9.png)

# Token count

![image](https://user-images.githubusercontent.com/1238659/70188067-93e61b00-16a4-11ea-87d9-5d44bcf4d4e4.png)

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

cjcenizal commented:
@elasticmachine merge upstream

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13334/)
* Commit: e008b54328a226c97ce60ce7ca59029b03207846

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13345/)
* Commit: 8d03daeee69d936ba84a410536a9b70323574bf1

cjcenizal commented:
@elasticmachine merge upstream

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13471/)
* Commit: 346ea64efa6ede51157ebdcb4f17bb92a1a71577

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13470/)
* Commit: 47c31aa83cdc101ca915d0dde96ab7f511247b25

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13484/)
* Commit: eafda4b93279cc46c861d5237043bf7d8ba6f2a5

cjcenizal commented:
@alisonelizabeth Great catches! I combed through the UI, fixed the issues you noted, and a few others. Could you please take another look?

cjcenizal commented:
@elasticmachine merge upstream

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13588/)
* Commit: 43e72d2182a5cfbf47d671acc022e6bf0665ba9d

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13597/)
* Commit: e636a7e20f7363f47ec3a41b046624173b1ced71

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/fields/field_types/token_count_type.tsx:
```diff
@@ -58,6 +58,10 @@ export const TokenCountType = ({ field }: Props) => {
       <AdvancedSettingsWrapper>
         <EditFieldSection>
           {/* null_value */}
+          <NullValueParameter
```
was this a mistake? the `NullValueParameter` component is rendered twice now.

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/field_parameters/term_vector_parameter.tsx:
```diff
@@ -0,0 +1,71 @@
+                  <p>
+                    {i18n.translate('xpack.idxMgmt.mappingsEditor.termVectorFieldWarningMessage', {
+                      defaultMessage:
+                        'Setting "With positions offsets" will double the size of a field’s index.',
```
I think this should be "With positions & offsets" to match the select option

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/field_parameters/path_parameter.tsx:
```diff
@@ -90,6 +91,7 @@ export const PathParameter = ({ field, allFields }: Props) => {
               }
             )}
             idAria="mappingsEditorPathParameter"
```
I think this needs to be changed to `ariaId`

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/field_parameters/analyzers_parameter.tsx:
```diff
@@ -0,0 +1,94 @@
+              <EuiSpacer />
+              <AnalyzerParameter
+                path="search_analyzer"
+                defaultValue={field.source.searchAnalyzer}
```
should this be `field.source.search_analyzer`?

cjcenizal commented:
Thanks @alisonelizabeth! I fixed the things you caught. I reverted my change to the way the analyzers were tracking the `useSameAnalyzerForSearch` state. This was originally an attempt to fix https://github.com/elastic/kibana/issues/52357.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13842/)
* Commit: 789370a99781a5e1f96ca502f6b80ffa56a99b6a

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/constants/parameters_definition.ts:
```diff
@@ -168,10 +175,48 @@ export const PARAMETERS_DEFINITION = {
+  },
+  null_value_boolean: {
+    fieldConfig: {
+      defaultValue: '',
```
I _think_ we should probably set a `defaultValue` here to one of the available options (e.g., `'true'`).

cjcenizal replied:
Thanks for pointing this out! According to the docs:

> Accepts any of the true or false values listed above. The value is substituted for any explicit null values. Defaults to null, which means the field is treated as missing.

That last bit is what trips me up. It defaults to null, so maybe we need to add that as an option and default to that option? I will merge this PR now so we can build off it without introducing merge conflicts and then investigate and address all of the various null default values for the different data types in a subsequent PR.

alisonelizabeth replied:
Sounds good!

Yeah, it is a little confusing. 

I don't think it's valid to set it to `null`.  If I try that, I get an error from ES: `Property [null_value] cannot be null.` I think the default behavior is for it not to be set. `boolean` is a little different from some of the other types that have this parameter in that it only excepts a set number of values, that is why we decided to make it a select instead of a text field. By setting `defaultValue` to an empty string, if the user enables the switch, but doesn't change the value from the dropdown, no value is sent back, even though it appears `"true"` is selected because it is the first item in the list of options. Maybe we should create an empty first option so the user is forced to make a selection?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13850/)
* Commit: 10376bdd71fbabf5701d2de5030c085d4f5f7613

sebelga commented on x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/components/document_fields/field_parameters/analyzers_parameter.tsx:
```diff
@@ -0,0 +1,90 @@
+              defaultMessage: 'Use the same analyzers for index and searching',
+            }
+          ),
+          defaultValue: true,
```
@cjcenizal This is what I tried to avoid, defining default values in multiple places in the code. I'd prefer to have a single source of truth for the parameter definition (their label, default value, ...). There is a helper function that we use for that: `getFieldConfig(<parameter>)`
